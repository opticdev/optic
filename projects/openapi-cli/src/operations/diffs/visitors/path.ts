import { OpenAPIV3 } from '../../../specs';
import { OperationDiffResult, OperationDiffResultKind } from '../result';
import { PathComponents, PathComponent } from '../..';
import { Option } from 'ts-results';

export function* visitPath(
  path: string,
  specOption: Option<OpenAPIV3.PathItemObject>,
  context: { pathPattern: Option<string> }
): IterableIterator<OperationDiffResult> {
  if (specOption.none) {
    yield {
      kind: OperationDiffResultKind.UnmatchedPath,
      subject: path,
    };
  }

  // TODO: match path's parameters against those documented
  // let parameterComponents = PathComponents.fromPath(path).filter(
  //   PathComponent.isTemplate
  // );
  // const spec = specOption.unwrap()
  // spec.parameters
}
