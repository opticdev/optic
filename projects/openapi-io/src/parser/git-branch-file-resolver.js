'use strict';
const fs = require('fs');
const { ono } = require('@jsdevtools/ono');
const url = require('@apidevtools/json-schema-ref-parser/lib/util/url');
const {
  ResolverError,
} = require('@apidevtools/json-schema-ref-parser/lib/util/errors');
const exec = require('child_process').exec;
const path = require('path');

module.exports = (gitBaseRepo, branch) => ({
  /**
   * The order that this resolver will run, in relation to other resolvers.
   *
   * @type {number}
   */
  order: 1,

  /**
   * Determines whether this resolver can read a given file reference.
   * Resolvers that return true will be tried, in order, until one successfully resolves the file.
   * Resolvers that return false will not be given a chance to resolve the file.
   *
   * @param {object} file           - An object containing information about the referenced file
   * @param {string} file.url       - The full URL of the referenced file
   * @param {string} file.extension - The lowercased file extension (e.g. ".txt", ".html", etc.)
   * @returns {boolean}
   */
  canRead(file) {
    return url.isFileSystemPath(file.url);
  },

  /**
   * Reads the given file and returns its raw contents as a Buffer.
   *
   * @param {object} file           - An object containing information about the referenced file
   * @param {string} file.url       - The full URL of the referenced file
   * @param {string} file.extension - The lowercased file extension (e.g. ".txt", ".html", etc.)
   * @returns {Promise<Buffer>}
   */
  read(file) {
    return new Promise((resolve, reject) => {
      const toGit = path.relative(gitBaseRepo, file.url);
      const command = `git show ${branch}:${toGit}`;
      try {
        exec(
          command,
          { cwd: gitBaseRepo, maxBuffer: 1024 * 1024 * 100 },
          (err, stdout, stderr) => {
            if (err)
              reject(
                new ResolverError(
                  ono(err.message, `Error opening file "${path}"`),
                  path
                )
              );
            if (stdout) resolve(stdout);
          }
        );
      } catch (err) {
        reject(
          new ResolverError(ono(err, `Error opening file "${path}"`), path)
        );
      }
    });
  },
});
