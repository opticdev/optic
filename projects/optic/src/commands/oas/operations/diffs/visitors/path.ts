import { OpenAPIV3 } from '../../../specs';
import {
  OperationDiffResult,
  OperationDiffResultKind,
} from '../../../../capture/patches/patchers/spec/types';
import { Option } from 'ts-results';

export function* visitPath(
  path: string,
  specOption: Option<OpenAPIV3.PathItemObject>,
  context: { pathPattern: Option<string> }
): IterableIterator<OperationDiffResult> {
  if (specOption.none) {
    yield {
      kind: OperationDiffResultKind.UnmatchedPath,
      subject: path,
    };
  }

  // TODO: match path's parameters against those documented
  // let parameterComponents = PathComponents.fromPath(path).filter(
  //   PathComponent.isTemplate
  // );
  // const spec = specOption.unwrap()
  // spec.parameters
}
