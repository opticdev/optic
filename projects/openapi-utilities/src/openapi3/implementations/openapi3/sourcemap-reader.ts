import * as YAML from 'yaml-ast-parser';
import {
  Kind,
  YamlMap,
  YAMLNode,
  YAMLSequence,
  YAMLMapping,
} from 'yaml-ast-parser';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import {
  ILookupFileResult,
  ILookupPathResult,
  JsonPath,
  SerializedSourcemap,
  SourcemapLine,
  ToSource,
} from './types';
import isUrl from 'is-url';
import urljoin from 'url-join';
import chalk from 'chalk';
import path from 'path';

export function resolveJsonPointerInYamlAst(
  node: YAMLNode, // root ast
  pointer: string
): YAMLNode | undefined {
  const decoded = jsonPointerHelpers.decode(pointer);
  const isEmpty =
    decoded.length === 0 || (decoded.length === 1 && decoded[0] === '');

  if (isEmpty) return node;

  const found: YAMLNode | undefined = decoded.reduce(
    (current, path) => {
      if (!current) return undefined;
      const node: YAMLNode = current.key ? current.value : current;
      const isNumericalKey =
        !isNaN(Number(path)) && (node as any).hasOwnProperty('items');

      if (isNumericalKey) {
        return (node as YAMLSequence).items[Number(path)];
      } else {
        const field = node.mappings.find(
          (i: YAMLMapping) => i.key.value === path
        );
        return field;
      }
    },
    node as YAMLNode | undefined
  );

  return found;
}

export function sourcemapReader(sourcemap: SerializedSourcemap) {
  const filesToYamlNode: Record<string, YAMLNode> = sourcemap.files.reduce(
    (acc, file) => ({
      ...acc,
      [file.path]: YAML.safeLoad(file.contents),
    }),
    {}
  );
  const rootFileNumber = sourcemap.files.find(
    (i) => i.path === sourcemap.rootFilePath
  )!.index;

  const findFile = (
    jsonPathFromRoot: JsonPath
  ): ILookupPathResult | undefined => {
    const fileResult = findFilePosition(jsonPathFromRoot);
    if (!fileResult) return undefined;

    const file = sourcemap.files.find((i) => i.path === fileResult.filePath)!;

    const node = resolveJsonPointerInYamlAst(
      filesToYamlNode[file.path],
      fileResult.startsAt
    );

    if (node)
      return {
        filePath: file.path,
        astNode: node,
        contents: file.contents,
        startsAt: fileResult.startsAt,
      };
  };

  const findFilePosition = (jsonPathFromRoot: JsonPath): ILookupFileResult => {
    const decoded = jsonPointerHelpers.decode(jsonPathFromRoot);

    let cursor: {
      currentFile: number;
      pathInRoot: string[];
      pathInCurrentFile: string[];
    } = {
      currentFile: rootFileNumber,
      pathInRoot: [],
      pathInCurrentFile: [],
    };

    decoded.forEach((component) => {
      const path = jsonPointerHelpers.compile([
        ...cursor.pathInCurrentFile,
        component,
      ]);

      cursor.pathInRoot.push(component);
      const hitRef = sourcemap.refMappings[path] as ToSource | undefined;

      if (hitRef) {
        const [file, startingPath] = hitRef;
        cursor.currentFile = file;
        cursor.pathInCurrentFile = jsonPointerHelpers.decode(startingPath);
      } else {
        cursor.pathInCurrentFile.push(component);
      }
    });

    const file = sourcemap.files.find((i) => i.index === cursor.currentFile)!;

    const cursorPath =
      cursor.pathInCurrentFile[0] === ''
        ? cursor.pathInCurrentFile.slice(1)
        : cursor.pathInCurrentFile;
    const pathInFile = jsonPointerHelpers.compile(cursorPath);

    return {
      filePath: file.path,
      startsAt: pathInFile,
    };
  };

  const findFileAndLines = (
    jsonPathFromRoot: JsonPath
  ): SourcemapLine | undefined => {
    const lookupResult = findFile(jsonPathFromRoot);
    if (lookupResult) {
      const astNode = lookupResult.astNode;
      const contents = lookupResult.contents;

      const [startPosition, endPosition] = astNodesToStartEndPosition(astNode);

      const { startLine, endLine } = positionToLine(
        contents,
        startPosition,
        endPosition
      );
      const result = {
        filePath: lookupResult.filePath,
        startLine,
        endLine,
        startPosition: startPosition,
        endPosition: endPosition,
      };
      return result;
    }
  };

  const findLinesForAstAndContents = (astNode: YAMLNode, contents: string) => {
    const [startPosition, endPosition] = astNodesToStartEndPosition(astNode);

    const { startLine, endLine } = positionToLine(
      contents,
      startPosition,
      endPosition
    );
    const result = {
      startLine,
      endLine,
      startPosition: startPosition,
      endPosition: endPosition,
    };
    return result;
  };

  return {
    findFile,
    findFilePosition,
    findFileAndLines,
    findLinesForAstAndContents,
  };
}

//////////////////////////////////////////////////////////

function positionToLine(
  contents: string,
  start: number,
  end: number
): { startLine: number; endLine: number; preview: string } {
  const startLine =
    (contents.substring(0, start).match(/\n/g) || '').length + 1;
  const endLine =
    (contents.substring(start, end).match(/\n/g) || '').length + startLine;

  const lines = contents.split(/\r\n|\r|\n/);

  const preview = lines.slice(startLine - 1, endLine).join('\n');

  return {
    startLine,
    endLine,
    preview,
  };
}

function astNodesToStartEndPosition(astNode: YAMLNode): [number, number] {
  try {
    switch (astNode.kind) {
      case Kind.MAP: {
        const map = astNode as YamlMap;
        const end =
          map.value.mappings[map.value.mappings.length - 1]?.endPosition ||
          astNode.endPosition;
        return [map.startPosition, end];
      }
      case Kind.SEQ: {
        const seq = astNode as YAMLSequence;
        const end =
          seq.items[seq.items.length - 1]?.endPosition || astNode.endPosition;
        return [seq.startPosition, end];
      }
      default:
        return [astNode.startPosition, astNode.endPosition];
    }
  } catch {
    return [astNode.startPosition, astNode.endPosition];
  }
}

export type GetSourcemapOptions =
  | {
      ciProvider: 'github' | 'gitlab';
      remote: string;
      sha: string;
      root: string;
    }
  | {
      ciProvider?: undefined;
    };

export function getSourcemapLink(
  sourcemap: SourcemapLine,
  options: GetSourcemapOptions = {}
): string {
  if (isUrl(sourcemap.filePath)) {
    return `${chalk.underline(sourcemap.filePath)} line ${sourcemap.startLine}`;
  } else if (options.ciProvider) {
    const pathFromRoot = path.relative(options.root, sourcemap.filePath);
    return options.ciProvider === 'github'
      ? urljoin(
          options.remote,
          'tree',
          options.sha,
          `${pathFromRoot}#L${sourcemap.startLine}`
        )
      : urljoin(
          options.remote,
          '-/blob',
          options.sha,
          `${pathFromRoot}#L${sourcemap.startLine}`
        );
  } else {
    const relativePath = path.relative(process.cwd(), sourcemap.filePath);
    return chalk.underline(
      `${relativePath}:${sourcemap.startLine}:${sourcemap.startPosition}`
    );
  }
}
