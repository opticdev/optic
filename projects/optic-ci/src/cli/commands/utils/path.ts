import path from 'path';
import { promisify } from 'util';
import { exec as callbackExec } from 'child_process';
const exec = promisify(callbackExec);

export const getRelativeRepoPath = (
  relativePath: string,
  gitRootPath: string | false
): string => {
  const absolutePath = path.join(process.cwd(), relativePath);

  if (!gitRootPath) {
    return absolutePath;
  } else {
    return absolutePath.startsWith(gitRootPath)
      ? absolutePath.replace(gitRootPath, '')
      : absolutePath;
  }
};

export const getGitRootPath = async () => {
  const { stdout: gitRootUntrimmed } = await exec(
    'git rev-parse --show-toplevel'
  );
  const gitRoot = gitRootUntrimmed.trim();
  return gitRoot;
};
