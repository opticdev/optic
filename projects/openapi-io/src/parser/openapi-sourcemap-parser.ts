import $RefParser from '@apidevtools/json-schema-ref-parser';
// @ts-ignore
import * as $RefParserOptions from '@apidevtools/json-schema-ref-parser/lib/options';
// @ts-ignore
import { dereference } from './insourced-dereference';
import path from 'path';

import fetch from 'node-fetch';
import isUrl from 'is-url';

import { JsonSchemaSourcemap } from './sourcemap';
import { gitBranchResolver } from './resolvers/git-branch-file-resolver';
import { ExternalRefHandler } from './types';
import Bottleneck from 'bottleneck';
// @ts-ignore
import * as customYaml from './insourced-yaml';
import {
  UserDefinedHeadersByUrlPrefix,
  customHttpResolver,
  parseHeadersConfig,
  DEFAULT_HEADERS,
  getMostRelevantHeader,
} from './resolvers/custom-http-ref-handler';
import {
  FlatOpenAPIV2,
  FlatOpenAPIV3,
  FlatOpenAPIV3_1,
} from '@useoptic/openapi-utilities';

export {
  JSONParserError,
  ResolverError,
} from '@apidevtools/json-schema-ref-parser';

export type ParseOpenAPIResult<
  T extends
    | FlatOpenAPIV2.Document
    | FlatOpenAPIV3.Document
    | FlatOpenAPIV3_1.Document,
> = {
  jsonLike: T;
  sourcemap: JsonSchemaSourcemap;
};

export async function dereferenceOpenApi<
  T extends
    | FlatOpenAPIV2.Document
    | FlatOpenAPIV3.Document
    | FlatOpenAPIV3_1.Document,
>(
  path: string,
  options: {
    externalRefHandler?: ExternalRefHandler;
    externalRefHeaders?: UserDefinedHeadersByUrlPrefix;
  } = {}
): Promise<ParseOpenAPIResult<T>> {
  const resolver = new $RefParser();
  const headersMap = parseHeadersConfig(options.externalRefHeaders ?? []);

  const sourcemap = new JsonSchemaSourcemap(path);
  const resolve = {
    file: options.externalRefHandler,
    customRefHandler: options.externalRefHandler,
    customHttpRefHandler: customHttpResolver(headersMap),
    external: true,
  };
  // Resolve all references
  const resolverResults: $RefParser.$Refs = await resolver.resolve(path, {
    resolve,
    parse: {
      yaml: customYaml,
    },
  });

  const limiter = new Bottleneck({
    maxConcurrent: 20,
  });

  // parse all asts and add to sourcemap
  const cachedUrls = new Set<string>([]);

  await Promise.all(
    resolverResults.paths().map((filePath, index) =>
      limiter.schedule(async () => {
        if (isUrl(filePath)) {
          const inCache = cachedUrls.has(filePath);
          if (!inCache) {
            const headers = {
              ...DEFAULT_HEADERS,
              ...getMostRelevantHeader(filePath, headersMap),
            };
            const response = await fetch(filePath, {
              headers,
            });
            const contents = await response.text();
            cachedUrls.add(filePath);
            sourcemap.addFileIfMissingFromContents(filePath, contents, index);
          }
        } else if (options.externalRefHandler) {
          const contents = await options.externalRefHandler.read({
            url: filePath,
          });
          sourcemap.addFileIfMissingFromContents(filePath, contents, index);
        } else {
          await sourcemap.addFileIfMissing(filePath, index);
        }
      })
    )
  );

  // Dereference all references
  dereference(
    resolver,
    {
      ...$RefParserOptions.defaults,
      path: path,
      dereference: { circular: 'ignore' },
      resolve,
      parse: {
        yaml: customYaml,
      },
    },
    sourcemap
  );

  return { jsonLike: resolver.schema as any, sourcemap: sourcemap };
}

export async function parseOpenAPIWithSourcemap<
  T extends
    | FlatOpenAPIV2.Document
    | FlatOpenAPIV3.Document
    | FlatOpenAPIV3_1.Document,
>(
  path: string,
  options: {
    externalRefHeaders?: UserDefinedHeadersByUrlPrefix;
  } = {}
): Promise<ParseOpenAPIResult<T>> {
  return dereferenceOpenApi(path, {
    externalRefHeaders: options.externalRefHeaders,
  });
}

export async function parseOpenAPIFromRepoWithSourcemap<
  T extends
    | FlatOpenAPIV2.Document
    | FlatOpenAPIV3.Document
    | FlatOpenAPIV3_1.Document,
>(
  name: string,
  repoPath: string,
  branch: string,
  options: {
    externalRefHeaders?: UserDefinedHeadersByUrlPrefix;
  } = {}
): Promise<ParseOpenAPIResult<T>> {
  const inGitResolver = gitBranchResolver(repoPath, branch);
  const fileName = path.join(repoPath, name);
  return dereferenceOpenApi(fileName, {
    externalRefHandler: inGitResolver,
    externalRefHeaders: options.externalRefHeaders,
  });
}
