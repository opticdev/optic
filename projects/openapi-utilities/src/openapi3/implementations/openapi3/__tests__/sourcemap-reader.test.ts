// TODO get rid of fixture import so we don't have circular references
import { parseOpenAPIWithSourcemap } from '../../../../../../openapi-io/src/parser/openapi-sourcemap-parser';
import { it, expect, describe } from '@jest/globals';
import path from 'path';
import { sourcemapReader } from '../sourcemap-reader';
const util = require('util');
const exec = util.promisify(require('child_process').exec);

const fixture = async () => {
  return await parseOpenAPIWithSourcemap(
    path.resolve(
      path.join(
        __dirname,
        '../../../../../../openapi-io/inputs/openapi3-with-references/external-multiple.yaml'
      )
    )
  );
};

async function stripCwd(file: string) {
  const { stdout } = await exec('git rev-parse --show-toplevel');
  return (
    file
      .replace(stdout.trim(), '')
      // strip out monorail prefixes
      .replace('/projects/openapi-workspaces', '')
  );
}

describe('reading sourcemaps', () => {
  it('can get file path of forign file for any pointer relative to flattened json', async () => {
    const results = await fixture();

    const node = sourcemapReader(results.sourcemap).findFile(
      '/properties/user/example/name'
    );
    expect(node?.filePath.endsWith('definitions.yaml')).toBeTruthy();
    expect(await stripCwd(node!.filePath)).toMatchSnapshot();
  });

  it('can get file paths json pointer in main file relative to flattened json', async () => {
    const results = await fixture();

    const node = sourcemapReader(results.sourcemap).findFile('/properties');

    expect(node!.filePath.endsWith('external-multiple.yaml')).toBeTruthy();
    expect(await stripCwd(node!.filePath)).toMatchSnapshot();
  });

  it('lines in affected file for sub-property', async () => {
    const results = await fixture();

    const node = await sourcemapReader(results.sourcemap).findFileAndLines(
      '/properties/user/example/name'
    );
    node!.filePath = await stripCwd(node!.filePath);
    expect(node).toMatchSnapshot();
  });

  it('lines in affected for object', async () => {
    const results = await fixture();

    const node = await sourcemapReader(results.sourcemap).findFileAndLines(
      '/properties'
    );

    node!.filePath = await stripCwd(node!.filePath);
    expect(node).toMatchSnapshot();
  });

  it('resolve a large sampling of known-valid keys', async () => {
    const results = await parseOpenAPIWithSourcemap(
      path.resolve(
        path.join(
          __dirname,
          '../../../../../../openapi-io/inputs/openapi3/001-ok-add-property-field.yaml'
        )
      )
    );

    const knownValid = [
      '/paths/~1thing~1{thing_id}/patch/responses/200/content/application~1vnd.api+json/schema/properties/data/properties/attributes/properties/description',
      '/paths/~1thing~1{thing_id}/patch/responses/200/content/application~1vnd.api+json/schema/properties/data/properties/attributes/properties/created',
      '/paths/~1thing~1{thing_id}/patch/responses/409/content/application~1vnd.api+json/schema/properties/jsonapi',
      '/paths/~1thing~1{thing_id}/patch/responses/400/content/application~1vnd.api+json/schema/properties/jsonapi',
      '/paths/~1thing/get/responses/404/content/application~1vnd.api+json/schema/properties/errors',
      '/paths/~1thing~1{thing_id}/delete/responses/401/content/application~1vnd.api+json/schema/properties/errors',
      '/paths/~1thing~1{thing_id}/delete/responses/403/content/application~1vnd.api+json/schema/properties/jsonapi/properties/version',
      '/paths/~1thing~1{thing_id}/get/responses/400/headers/snyk-version-served',
      '/paths/~1thing~1{thing_id}/patch/responses/404/content/application~1vnd.api+json/schema/properties/errors/items/properties/id',
      '/paths/~1thing~1{thing_id}/delete/responses/401/content/application~1vnd.api+json/schema/properties/jsonapi',
      '/paths/~1thing/get/responses/500/content/application~1vnd.api+json/schema/properties/errors',
      '/paths/~1thing/post/responses/409/content/application~1vnd.api+json/schema/properties/jsonapi',
      '/paths/~1thing~1{thing_id}/patch/responses/401/content/application~1vnd.api+json/schema/properties/errors/items/properties/status',
      '/paths/~1thing~1{thing_id}/get/responses/403/content/application~1vnd.api+json/schema/properties/errors/items/properties/source/properties/pointer',
      '/paths/~1thing/get/responses/200/content/application~1vnd.api+json/schema/properties/data/items/properties/attributes/properties/description',
      '/paths/~1thing~1{thing_id}/delete/responses/403/content/application~1vnd.api+json/schema/properties/errors/items/properties/detail',
      '/paths/~1thing~1{thing_id}/patch/responses/500/content/application~1vnd.api+json/schema/properties/errors/items/properties/source/properties/parameter',
      '/paths/~1thing/get/responses/200/content/application~1vnd.api+json/schema/properties/data/items/properties/relationships',
      '/paths/~1thing~1{thing_id}/get/responses/500/content/application~1vnd.api+json/schema/properties/errors/items/properties/id',
      '/paths/~1thing~1{thing_id}/patch/responses/404/content/application~1vnd.api+json/schema/properties/errors/items/properties/detail',
      '/paths/~1thing~1{thing_id}/patch/responses/500/content/application~1vnd.api+json/schema/properties/errors/items/properties/source',
      '/paths/~1thing~1{thing_id}/delete/responses/404/content/application~1vnd.api+json/schema/properties/errors/items/properties/id',
      '/paths/~1thing~1{thing_id}/get/responses/403/headers/snyk-version-served',
      '/paths/~1thing/post/responses/403/content/application~1vnd.api+json/schema/properties/errors/items/properties/source/properties/parameter',
      '/paths/~1thing/get/responses/404/content/application~1vnd.api+json/schema/properties/errors/items/properties/source/properties/pointer',
      '/paths/~1thing~1{thing_id}/get/responses/400/headers/snyk-request-id',
      '/paths/~1thing/post/responses/404/content/application~1vnd.api+json/schema/properties/errors/items/properties/detail',
      '/paths/~1thing~1{thing_id}/get/responses/500/content/application~1vnd.api+json/schema/properties/jsonapi/properties/version',
      '/paths/~1thing~1{thing_id}/get/responses/500/content/application~1vnd.api+json/schema/properties/errors',
      '/paths/~1thing~1{thing_id}/delete/responses/400/content/application~1vnd.api+json/schema/properties/jsonapi',
      '/paths/~1thing~1{thing_id}/patch/responses/401/content/application~1vnd.api+json/schema/properties/errors/items/properties/source',
      '/paths/~1thing/get/responses/200/content/application~1vnd.api+json/schema/properties/data/items/properties/relationships/properties/example',
      '/paths/~1thing~1{thing_id}/get/responses/200/content/application~1vnd.api+json/schema/properties/data/properties/relationships/properties/example/properties/links/properties/related',
      '/paths/~1thing~1{thing_id}/delete/responses/500/content/application~1vnd.api+json/schema/properties/errors/items/properties/status',
      '/paths/~1thing~1{thing_id}/patch/responses/401/content/application~1vnd.api+json/schema/properties/jsonapi/properties/version',
      '/paths/~1thing~1{thing_id}/delete/responses/500/headers/deprecation',
      '/paths/~1thing/get/responses/404/content/application~1vnd.api+json/schema/properties/errors/items/properties/id',
      '/paths/~1thing~1{thing_id}/patch/responses/403/headers/snyk-version-served',
      '/paths/~1thing~1{thing_id}/delete/responses/403/content/application~1vnd.api+json/schema/properties/errors/items/properties/meta',
      '/paths/~1thing~1{thing_id}/patch/responses/401/content/application~1vnd.api+json/schema/properties/errors/items/properties/source/properties/pointer',
      '/paths/~1thing~1{thing_id}/delete/responses/500/content/application~1vnd.api+json/schema/properties/errors/items/properties/source/properties/pointer',
      '/paths/~1thing~1{thing_id}/delete/responses/400/content/application~1vnd.api+json/schema/properties/errors/items/properties/id',
      '/paths/~1thing/get/responses/404/headers/deprecation',
      '/paths/~1thing~1{thing_id}/patch/responses/404/content/application~1vnd.api+json/schema/properties/errors/items/properties/detail',
      '/paths/~1thing/get/responses/400/headers/snyk-version-lifecycle-stage',
      '/paths/~1thing~1{thing_id}/get/responses/404/headers/snyk-request-id',
      '/paths/~1thing/post/responses/409/content/application~1vnd.api+json/schema/properties/errors/items/properties/source/properties/pointer',
      '/paths/~1thing~1{thing_id}/delete/responses/400/content/application~1vnd.api+json/schema/properties/errors/items/properties/id',
      '/paths/~1thing/post/responses/404/content/application~1vnd.api+json/schema/properties/errors',
      '/paths/~1thing~1{thing_id}/get/responses/401/content/application~1vnd.api+json/schema/properties/jsonapi/properties/version',
    ];

    //
    const allResolved = await Promise.all(
      knownValid.map(async (i) => {
        const a = await sourcemapReader(results.sourcemap).findFileAndLines(i);
        const lastIndex = a!.filePath.lastIndexOf('/');
        a!.filePath = a!.filePath.substring(lastIndex);
        return a;
      })
    );

    expect(allResolved).toMatchSnapshot();
    expect(allResolved.every((i) => Boolean(i)));
  });
});
