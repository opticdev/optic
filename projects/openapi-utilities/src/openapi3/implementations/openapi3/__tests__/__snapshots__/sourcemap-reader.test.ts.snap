// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reading sourcemaps can get file path of forign file for any pointer relative to flattened json 1`] = `"/projects/openapi-io/inputs/openapi3-with-references/definitions.yaml"`;

exports[`reading sourcemaps can get file paths json pointer in main file relative to flattened json 1`] = `"/projects/openapi-io/inputs/openapi3-with-references/external-multiple.yaml"`;

exports[`reading sourcemaps lines in affected file for sub-property 1`] = `
{
  "endLine": 8,
  "endPosition": 111,
  "filePath": "/projects/openapi-io/inputs/openapi3-with-references/definitions.yaml",
  "startLine": 8,
  "startPosition": 98,
}
`;

exports[`reading sourcemaps lines in affected for object 1`] = `
{
  "endLine": 7,
  "endPosition": 117,
  "filePath": "/projects/openapi-io/inputs/openapi3-with-references/external-multiple.yaml",
  "startLine": 3,
  "startPosition": 37,
}
`;

exports[`reading sourcemaps resolve a large sampling of known-valid keys 1`] = `
[
  {
    "endLine": 208,
    "endPosition": 9182,
    "filePath": "/001-ok-add-property-field.yaml",
    "startLine": 205,
    "startPosition": 9021,
  },
  {
    "endLine": 199,
    "endPosition": 8823,
    "filePath": "/001-ok-add-property-field.yaml",
    "startLine": 195,
    "startPosition": 8648,
  },
  {
    "endLine": 7,
    "endPosition": 123,
    "filePath": "/common.yaml",
    "startLine": 1,
    "startPosition": 0,
  },
  {
    "endLine": 7,
    "endPosition": 123,
    "filePath": "/common.yaml",
    "startLine": 1,
    "startPosition": 0,
  },
  {
    "endLine": 8,
    "endPosition": 172,
    "filePath": "/errors.yaml",
    "startLine": 5,
    "startPosition": 96,
  },
  {
    "endLine": 8,
    "endPosition": 172,
    "filePath": "/errors.yaml",
    "startLine": 5,
    "startPosition": 96,
  },
  {
    "endLine": 5,
    "endPosition": 69,
    "filePath": "/common.yaml",
    "startLine": 4,
    "startPosition": 42,
  },
  {
    "endLine": 7,
    "endPosition": 334,
    "filePath": "/headers.yaml",
    "startLine": 5,
    "startPosition": 168,
  },
  {
    "endLine": 17,
    "endPosition": 320,
    "filePath": "/errors.yaml",
    "startLine": 15,
    "startPosition": 279,
  },
  {
    "endLine": 7,
    "endPosition": 123,
    "filePath": "/common.yaml",
    "startLine": 1,
    "startPosition": 0,
  },
  {
    "endLine": 8,
    "endPosition": 172,
    "filePath": "/errors.yaml",
    "startLine": 5,
    "startPosition": 96,
  },
  {
    "endLine": 7,
    "endPosition": 123,
    "filePath": "/common.yaml",
    "startLine": 1,
    "startPosition": 0,
  },
  {
    "endLine": 19,
    "endPosition": 351,
    "filePath": "/errors.yaml",
    "startLine": 18,
    "startPosition": 325,
  },
  {
    "endLine": 26,
    "endPosition": 471,
    "filePath": "/errors.yaml",
    "startLine": 25,
    "startPosition": 440,
  },
  {
    "endLine": 208,
    "endPosition": 9182,
    "filePath": "/001-ok-add-property-field.yaml",
    "startLine": 205,
    "startPosition": 9021,
  },
  {
    "endLine": 21,
    "endPosition": 382,
    "filePath": "/errors.yaml",
    "startLine": 20,
    "startPosition": 356,
  },
  {
    "endLine": 28,
    "endPosition": 513,
    "filePath": "/errors.yaml",
    "startLine": 27,
    "startPosition": 480,
  },
  {
    "endLine": 182,
    "endPosition": 8363,
    "filePath": "/001-ok-add-property-field.yaml",
    "startLine": 178,
    "startPosition": 8205,
  },
  {
    "endLine": 17,
    "endPosition": 320,
    "filePath": "/errors.yaml",
    "startLine": 15,
    "startPosition": 279,
  },
  {
    "endLine": 21,
    "endPosition": 382,
    "filePath": "/errors.yaml",
    "startLine": 20,
    "startPosition": 356,
  },
  {
    "endLine": 29,
    "endPosition": 547,
    "filePath": "/errors.yaml",
    "startLine": 22,
    "startPosition": 387,
  },
  {
    "endLine": 17,
    "endPosition": 320,
    "filePath": "/errors.yaml",
    "startLine": 15,
    "startPosition": 279,
  },
  {
    "endLine": 7,
    "endPosition": 334,
    "filePath": "/headers.yaml",
    "startLine": 5,
    "startPosition": 168,
  },
  {
    "endLine": 28,
    "endPosition": 513,
    "filePath": "/errors.yaml",
    "startLine": 27,
    "startPosition": 480,
  },
  {
    "endLine": 26,
    "endPosition": 471,
    "filePath": "/errors.yaml",
    "startLine": 25,
    "startPosition": 440,
  },
  {
    "endLine": 50,
    "endPosition": 1527,
    "filePath": "/headers.yaml",
    "startLine": 44,
    "startPosition": 1294,
  },
  {
    "endLine": 21,
    "endPosition": 382,
    "filePath": "/errors.yaml",
    "startLine": 20,
    "startPosition": 356,
  },
  {
    "endLine": 5,
    "endPosition": 69,
    "filePath": "/common.yaml",
    "startLine": 4,
    "startPosition": 42,
  },
  {
    "endLine": 8,
    "endPosition": 172,
    "filePath": "/errors.yaml",
    "startLine": 5,
    "startPosition": 96,
  },
  {
    "endLine": 7,
    "endPosition": 123,
    "filePath": "/common.yaml",
    "startLine": 1,
    "startPosition": 0,
  },
  {
    "endLine": 29,
    "endPosition": 547,
    "filePath": "/errors.yaml",
    "startLine": 22,
    "startPosition": 387,
  },
  {
    "endLine": 70,
    "endPosition": 1447,
    "filePath": "/common.yaml",
    "startLine": 56,
    "startPosition": 1125,
  },
  {
    "endLine": 52,
    "endPosition": 1074,
    "filePath": "/common.yaml",
    "startLine": 41,
    "startPosition": 863,
  },
  {
    "endLine": 19,
    "endPosition": 351,
    "filePath": "/errors.yaml",
    "startLine": 18,
    "startPosition": 325,
  },
  {
    "endLine": 5,
    "endPosition": 69,
    "filePath": "/common.yaml",
    "startLine": 4,
    "startPosition": 42,
  },
  {
    "endLine": 31,
    "endPosition": 959,
    "filePath": "/headers.yaml",
    "startLine": 23,
    "startPosition": 662,
  },
  {
    "endLine": 17,
    "endPosition": 320,
    "filePath": "/errors.yaml",
    "startLine": 15,
    "startPosition": 279,
  },
  {
    "endLine": 7,
    "endPosition": 334,
    "filePath": "/headers.yaml",
    "startLine": 5,
    "startPosition": 168,
  },
  {
    "endLine": 32,
    "endPosition": 609,
    "filePath": "/errors.yaml",
    "startLine": 30,
    "startPosition": 552,
  },
  {
    "endLine": 26,
    "endPosition": 471,
    "filePath": "/errors.yaml",
    "startLine": 25,
    "startPosition": 440,
  },
  {
    "endLine": 26,
    "endPosition": 471,
    "filePath": "/errors.yaml",
    "startLine": 25,
    "startPosition": 440,
  },
  {
    "endLine": 17,
    "endPosition": 320,
    "filePath": "/errors.yaml",
    "startLine": 15,
    "startPosition": 279,
  },
  {
    "endLine": 31,
    "endPosition": 959,
    "filePath": "/headers.yaml",
    "startLine": 23,
    "startPosition": 662,
  },
  {
    "endLine": 21,
    "endPosition": 382,
    "filePath": "/errors.yaml",
    "startLine": 20,
    "startPosition": 356,
  },
  {
    "endLine": 23,
    "endPosition": 662,
    "filePath": "/headers.yaml",
    "startLine": 9,
    "startPosition": 336,
  },
  {
    "endLine": 50,
    "endPosition": 1527,
    "filePath": "/headers.yaml",
    "startLine": 44,
    "startPosition": 1294,
  },
  {
    "endLine": 26,
    "endPosition": 471,
    "filePath": "/errors.yaml",
    "startLine": 25,
    "startPosition": 440,
  },
  {
    "endLine": 17,
    "endPosition": 320,
    "filePath": "/errors.yaml",
    "startLine": 15,
    "startPosition": 279,
  },
  {
    "endLine": 8,
    "endPosition": 172,
    "filePath": "/errors.yaml",
    "startLine": 5,
    "startPosition": 96,
  },
  {
    "endLine": 5,
    "endPosition": 69,
    "filePath": "/common.yaml",
    "startLine": 4,
    "startPosition": 42,
  },
]
`;
